### PR TITLE
[C-Api] new API to flush the pipeline @open sesame 4/29 11:23

### PR DESCRIPTION
**Pull request by jaeyun-jung**
Add new API to flush the pipeline.
If new input should be run without old data in pipeline, this is necessary to clear all.

Signed-off-by: Jaeyun <jy1210.jung@samsung.com>

### CODE REVIEW

**Review by anyj0527 — APPROVED** [review]
Tested the new API and confirmed it on Android.
Thanks!

File: c/src/nnstreamer-capi-pipeline.c [hunk 1]
```diff
@@ -1076,6 +1076,38 @@ ml_pipeline_stop (ml_pipeline_h pipe)
...
+  status = ml_pipeline_stop (pipe);
+  if (status != ML_ERROR_NONE)
+    return status;
+
+  ml_logi ("The pipeline is stopped, clear all data from the pipeline.");
+
+  /* send flush event to pipeline */
+  g_mutex_lock (&p->lock);
+  gst_element_send_event (p->element, gst_event_new_flush_start ());
+  gst_element_send_event (p->element, gst_event_new_flush_stop (TRUE));
```
Comment (myungjoo) [inline c/src/nnstreamer-capi-pipeline.c:1102]: Don't you need to handle return values of ```gst_element_send_event()```?

Comment (jaeyun-jung) [reply c/src/nnstreamer-capi-pipeline.c]: updated

**Comment by jaeyun-jung** [comment]
2021-04-06 add testcase for pipeline flush (need nnstreamer PR https://github.com/nnstreamer/nnstreamer/pull/3179)

**Review by myungjoo — APPROVED** [review]
LGTM

File: c/src/nnstreamer-capi-pipeline.c [hunk 2]
```diff
@@ -1076,6 +1076,41 @@ ml_pipeline_stop (ml_pipeline_h pipe)
...
+
+  status = ml_pipeline_stop (pipe);
+  if (status != ML_ERROR_NONE)
+    return status;
+
+  ml_logi ("The pipeline is stopped, clear all data from the pipeline.");
+
+  /* send flush event to pipeline */
+  g_mutex_lock (&p->lock);
+  if (!gst_element_send_event (p->element, gst_event_new_flush_start ()) ||
```
Comment (anyj0527) [inline c/src/nnstreamer-capi-pipeline.c:1101]: Error report:
There is a case when tensor_mux fails to `flush_start`
Might need to fix tensor_mux

example log:
W/GStreamer+collectpads: 0:00:06.914059007 0x79663112d0 ../libs/gst/base/gstcollectpads.c:1711:gst_collect_pads_event_default:<mux:sink_0> forwarding flush start failed
E/GStreamer+queue: 0:00:06.914213434 0x79663112d0 ../plugins/elements/gstqueue.c:1028:gst_queue_handle_sink_event:<queue0> Failed to push event
W/GStreamer+collectpads: 0:00:06.914267341 0x79663112d0 ../libs/gst/base/gstcollectpads.c:1711:gst_collect_pads_event_default:<mux:sink_1> forwarding flush start failed
E/GStreamer+queue: 0:00:06.914379216 0x79663112d0 ../plugins/elements/gstqueue.c:1028:gst_queue_handle_sink_event:<queue1> Failed to push event
W/GStreamer+collectpads: 0:00:06.914428226 0x79663112d0 ../libs/gst/base/gstcollectpads.c:1711:gst_collect_pads_event_default:<mux:sink_2> forwarding flush start failed
E/GStreamer+queue: 0:00:06.914507705 0x79663112d0

**Review by anyj0527 — CHANGES_REQUESTED** [review]

File: c/src/nnstreamer-capi-pipeline.c (hunk 2 above)
Comment (jaeyun-jung) [reply c/src/nnstreamer-capi-pipeline.c:1101]: We cannot assure that all elements in pipeline successfully handle flush event.
I will change this condition and print warning message.

**Review by anyj0527 — APPROVED** [review]
Cool feature LGTM!

**Pull request merged**